# Projeto 2 - Classificador Automático de Sentimento

Você foi contratado por uma empresa parar analisar como os clientes estão reagindo a um determinado produto no Twitter. A empresa deseja que você crie um programa que irá analisar as mensagens disponíveis e classificará como "relevante" ou "irrelevante". Com isso ela deseja que mensagens negativas, que denigrem o nome do produto, ou que mereçam destaque, disparem um foco de atenção da área de marketing.<br /><br />
Como aluno de Ciência dos Dados, você lembrou do Teorema de Bayes, mais especificamente do Classificador Naive-Bayes, que é largamente utilizado em filtros anti-spam de e-mails. O classificador permite calcular qual a probabilidade de uma mensagem ser relevante dadas as palavras em seu conteúdo.<br /><br />
Para realizar o MVP (*minimum viable product*) do projeto, você precisa implementar uma versão do classificador que "aprende" o que é relevante com uma base de treinamento e compara a performance dos resultados com uma base de testes.<br /><br />
Após validado, o seu protótipo poderá também capturar e classificar automaticamente as mensagens da plataforma.

## Informações do Projeto

Prazo: 13/Set até às 23:59.<br />
Grupo: 1 ou 2 pessoas.<br /><br />
Entregáveis via GitHub: 
* Arquivo notebook com o código do classificador, seguindo as orientações abaixo.
* Arquivo Excel com as bases de treinamento e teste totalmente classificado.

**NÃO disponibilizar o arquivo com os *access keys/tokens* do Twitter.**


### Check 3: 

Até o dia 06 de Setembro às 23:59, o notebook e o xlsx devem estar no Github com as seguintes evidências: 
    * Conta no twitter criada.
    * Produto escolhido.
    * Arquivo Excel contendo a base de treinamento e teste já classificado.

Sugestão de leitura:<br />
http://docs.tweepy.org/en/v3.5.0/index.html<br />
https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/

___

## Preparando o ambiente

Instalando a biblioteca *tweepy* para realizar a conexão com o Twitter:

In [1]:
%%capture

#Instalando o tweepy
!pip install tweepy

Importando as Bibliotecas que serão utilizadas. Esteja livre para adicionar outras.

In [2]:
import tweepy
import math
import os.path
import pandas as pd
import json
from random import shuffle
from operator import mul
from functools import reduce

___
## Autenticando no  Twitter

Para realizar a captura dos dados é necessário ter uma conta cadastrada no twitter:

* Conta: ***[Preencha aqui o id da sua conta. Ex: @fulano ]***


1. Caso ainda não tenha uma: https://twitter.com/signup
1. Depois é necessário registrar um app para usar a biblioteca: https://apps.twitter.com/
1. Dentro do registro do App, na aba Keys and Access Tokens, anotar os seguintes campos:
    1. Consumer Key (API Key)
    1. Consumer Secret (API Secret)
1. Mais abaixo, gere um Token e anote também:
    1. Access Token
    1. Access Token Secret
    
1. Preencha os valores no arquivo "auth.pass"

**ATENÇÃO**: Nunca divulgue os dados desse arquivo online (GitHub, etc). Ele contém as chaves necessárias para realizar as operações no twitter de forma automática e portanto é equivalente a ser "hackeado". De posse desses dados, pessoas mal intencionadas podem fazer todas as operações manuais (tweetar, seguir, bloquear/desbloquear, listar os seguidores, etc). Para efeito do projeto, esse arquivo não precisa ser entregue!!!

In [3]:
#Dados de autenticação do twitter:

#Coloque aqui o identificador da conta no twitter: @fulano

#leitura do arquivo no formato JSON
with open('auth.pass') as fp:    
    data = json.load(fp)

#Configurando a biblioteca. Não modificar
auth = tweepy.OAuthHandler(data['consumer_key'], data['consumer_secret'])
auth.set_access_token(data['access_token'], data['access_token_secret'])

___
## Coletando Dados

Agora vamos coletar os dados. Tenha em mente que dependendo do produto escolhido, não haverá uma quantidade significativa de mensagens, ou ainda poder haver muitos retweets.<br /><br /> 
Configurando:

In [4]:
#Produto escolhido:
produto = 'renner'

#Quantidade mínima de mensagens capturadas:
n = 500
#Quantidade mínima de mensagens para a base de treinamento:
t = 300

#Filtro de língua, escolha uma na tabela ISO 639-1.
lang = 'pt'

Capturando os dados do twitter:

In [5]:
#Cria um objeto para a captura
api = tweepy.API(auth)

#Inicia a captura, para mais detalhes: ver a documentação do tweepy
i = 1
msgs = []
for msg in tweepy.Cursor(api.search, q=produto, lang=lang).items():    
    msgs.append(msg.text.lower())
    i += 1
    if i > n:
        break

#Embaralhando as mensagens para reduzir um possível viés
shuffle(msgs)

Salvando os dados em uma planilha Excel:

In [6]:
#Verifica se o arquivo não existe para não substituir um conjunto pronto
if not os.path.isfile('./{0}.xlsx'.format(produto)):
    
    #Abre o arquivo para escrita
    writer = pd.ExcelWriter('{0}.xlsx'.format(produto))

    #divide o conjunto de mensagens em duas planilhas
    dft = pd.DataFrame({'Treinamento' : pd.Series(msgs[:t])})
    dft.to_excel(excel_writer = writer, sheet_name = 'Treinamento', index = False)

    dfc = pd.DataFrame({'Teste' : pd.Series(msgs[t:])})
    dfc.to_excel(excel_writer = writer, sheet_name = 'Teste', index = False)

    #fecha o arquivo
    writer.save()

___
## Classificando as Mensagens

Agora você deve abrir o arquivo Excel com as mensagens capturadas e classificar na Coluna B se a mensagem é relevante ou não.<br /> 
Não se esqueça de colocar um nome para a coluna na célula **B1**.<br /><br />
Fazer o mesmo na planilha de Controle.

___
## Montando o Classificador Naive-Bayes

Com a base de treinamento montada, comece a desenvolver o classificador. Escreva o seu código abaixo:

Opcionalmente: 
* Limpar as mensagens removendo os caracteres: enter, :, ", ', (, ), etc. Não remover emojis.<br />
* Corrigir separação de espaços entre palavras e/ou emojis.
* Propor outras limpezas/transformações que não afetem a qualidade da informação.



In [7]:
excel=pd.ExcelFile("renner.xlsx")
excel.sheet_names
[u'Treinamento', u'Teste']
df = excel.parse("Treinamento")

df=df.replace({'@':''}, regex=True)
df=df.replace({'()':''}, regex=True)
df=df.replace({':':''}, regex=True)
df=df.replace({'#':''}, regex=True)
df=df.replace({';':''}, regex=True)
df=df.replace({'':''}, regex=True)
df=df.replace({',':''}, regex=True)
df=df.replace({'"':''}, regex=True)
df=df.replace({'  ':''}, regex=True)
df.head(10)



Treinamento  B1
0  tô procurando uma blusa rosa bebê encontro na ...   2
1  rt joanarborsa a pessoa elogia minha roupa e e...   1
2  menor aprendiz ipatinga 2017 renner vagas aten...   2
3  halfwayana deuradox eu vi na renner la em curi...   2
4  pedrlds vou comprar minhas brusinha da renner ...   1
5  rt joanarborsa a pessoa elogia minha roupa e e...   1
6     vi meu compadre na renner hj que comédia ele 💜   2
7                     doidoaltos viados nessa renner   2
8  i liked a youtube video https//t.co/rnkhebwhg5...   2
9  rt legalizandra vc paga o vestido em 3 prestaç...   1

In [8]:
relevante=[]
irrelevante=[]
total=[]
contandoRelevante={}
contandoIrrelevante={}
probPalavraRelevante={}
probPalavraIrrelevante={}

for i in range(len(df)):
    if df["B1"][i]==1:
        relevante+=(list(set(df['Treinamento'][i].split())))
        total+=(list(set(df['Treinamento'][i].split())))
    else:
        irrelevante+=(list(set(df['Treinamento'][i].split())))
        total+=(list(set(df['Treinamento'][i].split())))
        
nrelevante=len(relevante)
nirrelevante=len(irrelevante)
ntotal=len(total)

for i in relevante:
    total=relevante.count(i)
    contandoRelevante[i] = total

for i in irrelevante:
    total=irrelevante.count(i)
    contandoIrrelevante[i] = total

print(nirrelevante)
print(nrelevante)
print(ntotal)

2319
1948
4267


In [9]:
geral=[]
geral.extend(relevante)
geral.extend(irrelevante)

#Geral
geral_semRepeticao = set(geral)
numeroGeral_semRepeticao = len(geral_semRepeticao)

#Relevante
numeroRelevante_comRepeticao = len(relevante)
uRelevante_semRepeticao = set(relevante)
numeroRelevante_semRepeticao = len(uRelevante_semRepeticao)

#Irrelevante
numeroIrrelevante_comRepeticao = len(irrelevante)
uIrrelevante_semRepeticao = set(irrelevante)
numeroIrrelevante_semRepeticao = len(uIrrelevante_semRepeticao)

print(numeroGeral_semRepeticao)
print(numeroRelevante_semRepeticao)
print(numeroIrrelevante_semRepeticao)

1422
550
1111


In [10]:
for i in uRelevante_semRepeticao:
    probPalavraRelevante[i] = (contandoRelevante[i] + 1) / (numeroRelevante_comRepeticao + numeroGeral_semRepeticao)

for i in uIrrelevante_semRepeticao:
    probPalavraIrrelevante[i] = (contandoIrrelevante[i] + 1) / (numeroIrrelevante_comRepeticao + numeroGeral_semRepeticao)



In [11]:
dt = excel.parse("Teste")

dt=dt.replace({'@':''}, regex=True)
dt=dt.replace({'()':''}, regex=True)
dt=dt.replace({':':''}, regex=True)
dt=dt.replace({'#':''}, regex=True)
dt=dt.replace({';':''}, regex=True)
dt=dt.replace({'':''}, regex=True)
dt=dt.replace({',':''}, regex=True)
dt=dt.replace({'"':''}, regex=True)
dt=dt.replace({'  ':''}, regex=True)
dt.head(10)

Teste  B1
0                     mrsmckitchen liquidaçao renner   2
1  mário renner poluindo a minha tv que nojo fera...   2
2  vi uns tao lindos aqui na renner \nto querendo...   1
3  eu e a bia fomos na renner pra comprar minha c...   2
4  rt cecifangirling eu aceito uma brusinha da re...   1
5       meu irmão ta cantando rick e renner no banho   2
6  rt matheusanjos102 bgl é fazer cartão renner e...   1
7  rt legalizandra vc paga o vestido em 3 prestaç...   1
8  minhada renner não vai trabalhar sexta com que...   2
9                aquela que o rick e renner regravou   2

In [12]:
lteste=[]
prob_tweet = []
lprob_tweetrelevante = []
lprob_tweetirrelevante = []
relevantes_final = []

for i in dt.Teste:
    split_teste= i.split()
    lteste.append(split_teste)


for tweet in lteste:
    for palavra in tweet:
        if palavra in uRelevante_semRepeticao:
            prob_tweet.append(probPalavraRelevante[palavra])
        else:
            valor = 1 / (numeroRelevante_semRepeticao + numeroGeral_semRepeticao)
            prob_tweet.append(valor)
    lprob_tweetrelevante.append(reduce(mul,prob_tweet))
    del prob_tweet [:]

for tweet in lteste:
    for palavra in tweet:
        if palavra in uIrrelevante_semRepeticao:
            prob_tweet.append(probPalavraIrrelevante[palavra])
        else:
            valor = 1 / (numeroIrrelevante_semRepeticao + numeroGeral_semRepeticao)
            prob_tweet.append(valor)
    lprob_tweetirrelevante.append(reduce(mul,prob_tweet))
    del prob_tweet [:]
    
for i in range(len(lteste)):
    if lprob_tweetrelevante[i] > lprob_tweetirrelevante[i]:
        relevantes_final.append('Sim')
    if lprob_tweetrelevante[i] < lprob_tweetirrelevante[i]:
        relevantes_final.append('Não')
    if lprob_tweetrelevante[i] == lprob_tweetirrelevante[i]:
        relevantes_final.append('Neutro')

In [24]:
dt["Resultados"]=pd.Series(relevantes_final)
print(relevantes_final)

['Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Não', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Não', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Sim', 'Sim', 'Não', 'Sim', 'Sim', 'Não'

___
## Verificando a performance

Agora você deve testar o seu Classificador com a base de Testes.<br /><br /> 

Você deve extrair as seguintes medidas:
* Porcentagem de positivos falsos (marcados como relevante mas não são relevantes)
* Porcentagem de positivos verdadeiros (marcado como relevante e são relevantes)
* Porcentagem de negativos verdadeiros (marcado como não relevante e não são relevantes)
* Porcentagem de negativos falsos (marcado como não relevante e são relevantes)

Opcionalmente:
* Criar categorias intermediárias de relevância baseado na diferença de probabilidades. Exemplo: muito relevante, relevante, neutro, irrelevante e muito irrelevante.

In [36]:
listaResultados = []
for i in range(len(relevantes_final)):
    if dt.Resultados[i] == "Sim"  and dt.B1[i] == 1:
        listaResultados.append("Positivos verdadeiros")
    if dt.Resultados[i] == "Não"  and dt.B1[i] == 1:
        listaResultados.append("Negativos falsos")
    if dt.Resultados[i] == "Sim"  and dt.B1[i] == 2:
        listaResultados.append("Positivos falsos")
    if dt.Resultados[i] == "Não"  and dt.B1[i] == 2:
        listaResultados.append("Negativos verdadeiros")
        
dt["Final"] = pd.Series(listaResultados)

In [38]:
#Porcentagem de Negativos, Positivos, Falsos Postivos e Falsos Negativos
(dt.Final.value_counts(normalize=True)*100).round(decimals=1).reindex(['Positivos verdadeiros', 'Negativos falsos', 'Positivos falsos', 'Negativos verdadeiros'])

Positivos verdadeiros    37.0
Negativos falsos          1.5
Positivos falsos         45.0
Negativos verdadeiros    16.5
Name: Final, dtype: float64

___
## Concluindo

Escreva aqui a sua conclusão.<br /> 
Faça um comparativo qualitativo sobre as medidas obtidas.<br />
Explique como são tratadas as mensagens com dupla negação e sarcasmo.<br />
Proponha um plano de expansão. Por que eles devem continuar financiando o seu projeto?<br />

Opcionalmente: 
* Discorrer por que não posso alimentar minha base de Treinamento automaticamente usando o próprio classificador, aplicado a novos tweets.
* Propor diferentes cenários de uso para o classificador Naive-Bayes. Cenários sem intersecção com este projeto.
* Sugerir e explicar melhorias reais no classificador com indicações concretas de como implementar (não é preciso codificar, mas indicar como fazer e material de pesquisa sobre o assunto).


##### Decidimos escolher a loja de roupas Renner, pois é uma loja famosa e bem assecível para a população brasileira. Pode-se concluir que não obtivemos um resultado tão bom quanto esperávamos, pois nosso classificador apenas obteve sucesso em 53,5% dos casos, sendo esses, 37% de "Positivos Verdadeiros"(relevantes tanto pra nós quanto para o programa) e 16.5% de "Negativos verdadeiros"(irrelevante para nós e para o programa), ou seja, nosso classificador não obteve muito êxito classificando tweets "Positivos", pois os "Positivos Falsos"(Relevante para o programa e não para nós) representaram 45% .

##### Em relação aos tweets sarcásticos e com dupla negação, nosso classificador, por ser baseado no teorema de Naive Bayes, não é capaz de reconhecer sarcasmo, uma vez que esse teorema é muito simples e trabalha apenas calculando probabilidade de palavras, e não de frases, e geralmente o sarcasmo só pode ser identificado ao se ler a frase toda.   

##### Um plano de expansão possível seria filtrar mais comentários manualmente, para "ensinar" melhor o classificador, pois mesmo com 300 tweets de treinamento, muitas palavras se repetiram, tanto nos comentários relevantes quanto os irrelevantes. Também podemos separar por "tipos de roupa" e "preço" os comentários, pois diversos cometários falam mal do preço e diversos falam bem da roupa, logo, se separamos os dois, o classificador será mais preciso. Também podemos remover os comentários que são "rt" (retweets), pois os rt's aumentam a probabilidade das palavras que estão nele de serem relevantes ou irrelevantes, uma vez que são comentários exatemente iguais e repetidos. Eles devem continuar financiando nosso projeto pois este projeto, pode dar dicas de como eles podem atrair mais clientes (reajustando o preço de algud produtos, ou produzindo mais peças que estão em falta).  